pkg/tsdb/azuremonitor/httpclient.go

```diff
@@ -8,34 +8,38 @@ import (
 	"github.com/grafana/grafana/pkg/tsdb/azuremonitor/aztokenprovider"
 )
 
-func httpClientProvider(route azRoute, model datasourceInfo, cfg *setting.Cfg) (*httpclient.Provider, error) {
-	var clientProvider *httpclient.Provider
+func getMiddlewares(route azRoute, model datasourceInfo, cfg *setting.Cfg) ([]httpclient.Middleware, error) {
+	middlewares := []httpclient.Middleware{}
 
 	if len(route.Scopes) > 0 {
 		tokenProvider, err := aztokenprovider.NewAzureAccessTokenProvider(cfg, model.Credentials)
 		if err != nil {
 			return nil, err
 		}
+		middlewares = append(middlewares, aztokenprovider.AuthMiddleware(tokenProvider, route.Scopes))
+	}
 
-		clientProvider = httpclient.NewProvider(httpclient.ProviderOptions{
-			Middlewares: []httpclient.Middleware{
-				aztokenprovider.AuthMiddleware(tokenProvider, route.Scopes),
-			},
+	if _, ok := model.DecryptedSecureJSONData["appInsightsApiKey"]; ok {
+		// Inject API-Key for AppInsights
+		apiKeyMiddleware := httpclient.MiddlewareFunc(func(opts httpclient.Options, next http.RoundTripper) http.RoundTripper {
+			return httpclient.RoundTripperFunc(func(req *http.Request) (*http.Response, error) {
+				req.Header.Set("X-API-Key", model.DecryptedSecureJSONData["appInsightsApiKey"])
+				return next.RoundTrip(req)
+			})
 		})
-	} else {
-		clientProvider = httpclient.NewProvider()
+		middlewares = append(middlewares, apiKeyMiddleware)
 	}
 
-	return clientProvider, nil
+	return middlewares, nil
 }
 
-func newHTTPClient(route azRoute, model datasourceInfo, cfg *setting.Cfg) (*http.Client, error) {
+func newHTTPClient(route azRoute, model datasourceInfo, cfg *setting.Cfg, clientProvider httpclient.Provider) (*http.Client, error) {
 	model.HTTPCliOpts.Headers = route.Headers
-
-	clientProvider, err := httpClientProvider(route, model, cfg)
+	m, err := getMiddlewares(route, model, cfg)
 	if err != nil {
 		return nil, err
 	}
 
+	model.HTTPCliOpts.Middlewares = m
```

Comment:
This looks a bit scary. Is there a chance that you have one instance of `datasourceInfo` and `httpclient.Options` per configured Azure data source and you get a different middleware for different `route`? In that case you're basically modifying the state of `datasourceInfo` between requests and could be risk of nasty bugs.

Looking at the code calling this seems to be per each executor, but I'm not sure hence my comment above.